### DISCUSSION TITLE
GitHub/Vercel preview deployment integration

### DISCUSSION THREAD

**Discussion by 3x071c** [planetscale/discussion#175] [Feature Requests]
I'm working on a Next.js project that is currently being deployed on Vercel, so pairing it with PlanetScale felt intuitive when I first stumbled upon it. To understand my problem with PlanetScale atm, let's take a look at a common GitHub-centric CI/CD branching workflow ([Officially coined GitHub flow](https://docs.github.com/en/get-started/quickstart/github-flow)):

1. Dev creates a new source branch to work on an issue
2. Dev opens a PR to let others follow progress and give feedback
3. PR gets merged or closed eventually (and the branch with it)

This is how Vercel integrates very nicely with GitHub-centric workflows:

1. Dev creates a new source branch to work on an issue
    1. **Vercel automatically creates a development/preview deployment for all source changes**
2. Dev opens a PR to let others follow progress and give feedback
    1. **Vercel adds a status check to the PR, indicating a successful/failed deployment**, preventing a broken merge
3. PR gets merged or closed eventually (and the branch with it)
    1. **Vercel deploys the source changes from the development environment to production** if the PR passed

Adding Vercel (this can be extended to many Heroku-style deployment providers) to a project requires almost no configuration and no manual interference while developing. Everything just works as expected.
Here's how I initially expected PlanetScale to work:

1. Dev creates a new source branch to work on an issue
    1. Vercel automatically creates a development/preview deployment for all source changes
    2. **PlanetScale automatically creates a development branch for all schema changes**
3. Dev opens a PR to let others follow progress and give feedback
    1. Vercel adds a status check to the PR, indicating a successful/failed deployment, preventing a broken merge
    2. **PlanetScale adds a status check to the PR, indicating a schema conflict**, preventing a broken merge
4. PR gets merged or closed eventually (and the branch with it)
    1. Vercel deploys the source changes from the development environment to production if the PR passed
    2. **PlanetScale merges the schema changes from the development branch to production if the PR passed**

Right now, there's no such integration between GitHub and PlanetScale, which really makes things fall apart for developers like myself who prefer a code-first approach of using libraries like Prisma to modify a database schema. Creating a "Deploy Request" on PlanetScale, completely independent of the PR that made the schema change, is a horrible practice and not usable for production applications. If the PR doesn't merge and deploy at around the same time that the Deploy Request does, there's inconsistency between what the source code expects there to be in the database vs the actual database schema. This requires constant manual surveillance, intervention and effort to mirror all changes from GitHub to PlanetScale, which is not feasible.
Are there plans to address these concerns and create something like a PlanetScale GitHub integration, as an app or an action?

**Comment by tbarn** [comment]
Hi @3x071c! 

Thanks for writing this great feedback. It's definitely really helpful to hear how you would like to see the flow. This might also be helpful to share with the Vercel team, so I am going to do that internally too. In some ways it is not only a GitHub integration but also a deeper Vercel one that is needed. 

You are not the only one who has brought it up. I don't have a great answer for you right now, but just know that we are definitely thinking about it. Many of these flows don't feel new because we are used to Git flow, but for databases we are in new territory and we want to find how we can best support a wide range of development teams.

I've been personally exploring how to build GitHub Actions that interact with PlanetScale. You can see more in this post I wrote: https://planetscale.com/blog/using-the-planetscale-cli-with-github-actions-workflows It's not an out of the box solution, but I did it to explore where we could maybe go with it. I'd love to hear your thoughts on it beyond what you already share.

**Reply by 3x071c** [reply]
Hi, thanks for your reply! I too figured that my request is more about integrating Vercel's sandboxed/isolated preview deployment environments with a database such as PlanetScale (I changed the title accordingly). The problem really is that up until now, there's no convenient way to pair isolated database environments with every ephemeral Vercel preview deployment, and I think PlanetScale because of its branching feature is in a great position to take that spot. The linked blog post (+ [this similar project](https://github.com/jonico/pscale-workflow-helper-scripts)) was something I already had in my mind when I wrote this discussion post, and while GitHub Actions are definitely capable of doing this, writing those workflows manually (and the DX of actually using them in a project) isn't easy. I just tried the [PlanetScale Vercel integration](https://vercel.com/integrations/planetscale), which appears to lay the groundwork for providing separate environment variables/tokens for every temporary "preview" branch. 

For reference (also for others who are looking for a solution right now), [Prisma's Data Platform](https://www.prisma.io/dataplatform) is going to offer preview databases, and [Railway](https://railway.app) and Heroku already allow creating a separate environment with isolated resources for every PR (with their respective platforms drawbacks attached, of course). I'm eager to see how Vercel+PlanetScale will move along! 👍

**Reply by leerob** [reply]
Really appreciate your feedback here 🙏

**Reply by jcohenho** [reply]
@3x071c In the steps you describe for how you _expect_ the Planetscale integration to work, how are you currently handling the following to work with Vercel preview apps? 

> ii. PlanetScale automatically creates a development branch for all schema changes

Are you just manually copying the branch connection string and setting the preview DATABASE_URL env var? Curious if there's a way to automate attaching a Vercel preview app to a Planetscale branch

**Reply by sklum** [reply]
@jcohenho Are there branch-specific connection strings for PlanetScale?  I believe I'm trying to accomplish what you are as well based on your discussion [here](https://github.com/planetscale/discussion/discussions/187), and as far as I can tell the only connection strings available are to the main branch?

**Reply by tbarn** [reply]
@sklum Connection strings are branch specific. In the CLI or in the web app branch page, you will see a "Connect" button to create a connection string for that specific branch.

**Comment by JClackett** [comment]
Yeah this would be extremely useful, currently we have to:
- create a new git branch
- go to planetscale to create the new DB, and copy to .env file
- work on the feature
- publish branch
- create pr into main/whatever branch
- create deploy request for planetscale db
- go to vercel and add the DB url for that particular branch
- once pr is accepted, complete deploy request for planetscale, and merge pr

its not too too bad, but its a little annoying having the merge of the pr and the merge of the database be so disconnected, is there any better general practice for this? especially when going to production. On heroku, the release command for the build would run the prisma migration and if this failed the whole build would fail, meaning the DB + the actual production codebase remains on the previous succeeded build, allowing the dev to make whatever fixes are necessary and attempt again. However in planetscales scenario, how would you achieve something similar? ideally want to avoid having a database be updated, but then the build could fail and you're having to scramble to get the code building so that its in sync with the database again, and of course vice versa. Really forces you to think about backwards compatibility, which I guess is good :D

**Reply by mscoutermarsh** [reply]
Going to focus here specifically on the coupling of production code and database changes 😄 . And going into some depth since I believe will be helpful for others who come across this thread!

**Our recommendation:** Which comes from the experience of helping all kinds of business, from small to the absolute largest of scale. Is that in almost every case the schema change should be deployed first and should never be coupled with a code change. The last step you have in your flow described above is **perfect**.

**Why this flow is great:**
Having them separate forces best practices. Developers won't make risky code/schema changes because they can't 😄 . They must use the [expand/contract pattern](https://www.prisma.io/dataguide/types/relational/expand-and-contract-pattern) to safely make changes.

When the changes are coupled together, it always leads to some increased risk. It's impossible for code/schema to atomically deploy together, if they are ever dependent on each other going out at the same time, the app will error briefly until the other catches up.

As data size grows, migrations can take longer (think several minutes to days!), you don't want the app deployment blocked for this. 

Next, the case where something goes wrong with the schema change. When coupled together, then the deployment of the app is now blocked and a single change has stopped the pipeline of code going out into production until it's fixed.

**Places to improve:**
```
- go to planetscale to create the new DB, and copy to .env file
- work on the feature
- publish branch
- create pr into main/whatever branch
```

These steps right here are the time consuming part, and we're working on providing tooling to make these steps easy or even automatic.

For example, we recently released[ some tooling for Rails ](https://planetscale.com/blog/zero-downtime-rails-migrations-planetscale-rails-gem)which eliminates some of these steps + automates others. As well as a [collection of GitHub Actions](https://planetscale.com/blog/announcing-the-planetscale-github-actions).

I love the flow described above by @3x071c. I think the perfect workflow is one that automates the setup of the deploy request from a code change, but then leaves the developer to deploy the schema change, THEN deploy the code. That's the best of everything, easy on the developer + forces best practices that will serve the app well as it grows.

**Reply by JClackett** [reply]
Wow, thanks so much for taking the time to respond, and so quickly! And yeah, I can definitely see why automating the schema change into production can be dangerous. Safe to say you've convinced me! As you mentioned, being to able to automate the creation of the preview branches would be really nice and there's no risk in that side at all.

Again, thanks so much - super helpful to hear it from your side :)

**Reply by dani-mp** [reply]
A big improvement in my setup from the posted above is to have a constant database URL locally in your env var (e.g. `mysql://root@127.0.0.1:3306/my-db`) and then connect to it via the PlanetScale CLI (e.g. `pscale connect my-db my-branch`). The nice thing about this is that you can create a branch, connect to it and create a deploy request all from your terminal, without having to touch the local `.env` at all. You can easily switch branches too to test the expand and contract dance.

But yeah, the rest is not ideal. I usually forget to update my _preview_ branch (where my Vercel previews point to by default) whenever it gets obsolete. I think it should be stable as it is `main`. Also, for previews that require the newly deployed branch, I have to go to Vercel's dashboard and change the env var manually. Another part of the workflow I haven't solved yet either is to seed the new branches automatically whenever I create them.

In the beginning, I started using the GitHub actions but they were not robust and also didn't help me much, I might revisit this at some point.

**Reply by JClackett** [reply]
Yeahh that sounds good actually - and yeah going back and forth from vercel to local to planetscale is not ideal, I also forget to clean up old DATABASE_URLs, hard to keep track of what is connected to what.

**Reply by mattrossman** [reply]
[Neon's](https://neon.tech/) Vercel integration automatically sets up DB branches for preview deploys. 

https://vercel.com/integrations/neon

Just here to echo that it would helpful to have the same kind of integration for PlanetScale. Currently PlanetScale's Vercel integration only seems to set up the production branch.

**Comment by dakdevs** [comment]
> [Neon's](https://neon.tech/) Vercel integration automatically sets up DB branches for preview deploys.
> 
> https://vercel.com/integrations/neon
> 
> Just here to echo that it would helpful to have the same kind of integration for PlanetScale. Currently PlanetScale's Vercel integration only seems to set up the production branch.

This would be amazing!

**Comment by nwp** [comment]
This is the only reason I use Neon over PlanetScale. I would switch in a heartbeat if PS added this workflow like Neon does with Vercel.
